### PR TITLE
dialects: (dmp) cleanup

### PR DESCRIPTION
**Pull request by AntonLydike**
Renames:

 - `dmp.exchange_decl` -> `dmp.exchange`
 - `grid` attribute to `topo` on the `dmp.swap` operation.
 - Pass `dmp-decompose-2d` to `distribute-stencil`
 - `dmp.grid` to `dmp.topo`

Removals:
 - `ScatterOp`
 - `GatherOp`
 - Lowerings for both.
 - `HorizontalSlices2D` replaced by the generic slicing

### CODE REVIEW

File: xdsl/dialects/experimental/dmp.py [hunk 1]
```diff
@@ -160,7 +157,7 @@ def print_parameters(self, printer: Printer) -> None:
 
 
 @irdl_attr_definition
-class HaloShapeInformation(ParametrizedAttribute):
+class ShapeInformationAttr(ParametrizedAttribute):
```
Comment (georgebisbas) [inline xdsl/dialects/experimental/dmp.py:160]: ShapeAttr ?

Comment (AntonLydike) [reply xdsl/dialects/experimental/dmp.py]: Done!

**Review by georgebisbas — COMMENTED** [review]
Will have a review tmr morning!

**Comment by superlopuh** [comment]
@AntonLydike what's the plan for this PR?

File: xdsl/tools/command_line_tool.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def get_all_passes() -> list[type[ModulePass]]:
         convert_stencil_to_ll_mlir.ConvertStencilToLLMLIRPass,
         dead_code_elimination.DeadCodeElimination,
         DesymrefyPass,
-        stencil_global_to_local.GlobalStencilToLocalStencil2DHorizontal,
+        stencil_global_to_local.DistributeStencilPass,
```
Comment (georgebisbas) [inline xdsl/tools/command_line_tool.py:110]: Do we need the "Pass", should it be homogeneous across the rest?

File: xdsl/transforms/experimental/dmp/stencil_global_to_local.py [hunk 3]
```diff
@@ -636,39 +642,40 @@ class DmpDecompositionPass(ModulePass, ABC):
     Represents a pass that takes a strategy as input
     """
 
+
+@dataclass
+class DistributeStencilPass(DmpDecompositionPass):
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/dmp/stencil_global_to_local.py:647]: I know that we originally have this concept of "distributing a stencil", but maybe we should be smoothly starting to think on how to use the fact that we decompose the grid, not the stencil....
so like at some point to move to a decomposeGrid or sth similar Pass

File: xdsl/transforms/experimental/dmp/stencil_global_to_local.py [hunk 4]
```diff
@@ -636,39 +642,40 @@ class DmpDecompositionPass(ModulePass, ABC):
...
-    This pass *replaces* stencil shape inference in a
-    pass pipeline!
+    Name of the decomposition strategy to use, see STRATEGIES property for options
     """
 
-    name = "dmp-decompose-2d"
-
     restrict_domain: bool = True
+    """
+    Apply the domain restriction
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/dmp/stencil_global_to_local.py:672]: elaborate on what a restriction is?

File: tests/filecheck/dialects/dmp/canonicalize.mlir [hunk 5]
```diff
@@ -4,20 +4,20 @@ builtin.module {
     %ref = "test.op"() : () -> (memref<1024x1024xf32>)
 
     "dmp.swap"(%ref) {
-        grid = #dmp.grid<2x2>,
+        nodes = #dmp.grid<2x2>,
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/dialects/dmp/canonicalize.mlir:7]: Maybe we should call it ranks. The terminology around dmp would definitely benefit from standardization.

File: xdsl/dialects/experimental/dmp.py [hunk 6]
```diff
@@ -225,7 +222,7 @@ class HaloShapeInformation(ParametrizedAttribute):
         - halo_size(n, end  ) = dn - cn
     """
 
-    name = "dmp.shape_with_halo"
+    name = "dmp.shape"
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/dmp.py:225]: Yep!

File: xdsl/tools/command_line_tool.py (hunk 2 above)
Comment (AntonLydike) [reply xdsl/tools/command_line_tool.py:110]: Every class that is a pass should end in "...Pass", this makes discoverability in the codebase much better we found.

File: xdsl/transforms/experimental/dmp/stencil_global_to_local.py (hunk 3 above)
Comment (AntonLydike) [reply xdsl/transforms/experimental/dmp/stencil_global_to_local.py:647]: Yes! But I would wait with that for version 2 of the dmp dialect. It's definitely an interesting topic though!

File: xdsl/dialects/experimental/dmp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (georgebisbas) [reply xdsl/dialects/experimental/dmp.py:225]: SO why drop this from name? I guess it would help to have it around

**Review by georgebisbas — COMMENTED** [review]
@AntonLydike lemme know when you would like another pass

File: xdsl/dialects/experimental/dmp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/dmp.py:225]: I think at this level it's implied to have the halo. But I can add it back too

File: xdsl/dialects/experimental/dmp.py [hunk 7]
```diff
@@ -357,7 +354,7 @@ def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
 
 
 @irdl_attr_definition
-class NodeGrid(ParametrizedAttribute):
+class NodeGridAttr(ParametrizedAttribute):
     """
     This attribute specifies the node layout used to distribute the computation.
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/dmp.py:359]: Renamed to `RankTopoAttr`

File: tests/filecheck/dialects/dmp/canonicalize.mlir (hunk 5 above)
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/dialects/dmp/canonicalize.mlir]: Renamed it to `topo`

File: xdsl/dialects/experimental/dmp.py [hunk 8]
```diff
@@ -128,17 +125,17 @@ def elem_count(self) -> int:
         return prod(self.size)
 
     @property
-    def dim(self) -> int:
+    def rank(self) -> int:
         return len(self.size)
```
Comment (georgebisbas) [inline xdsl/dialects/experimental/dmp.py:129]: Why should a rank return len(self.size?)

File: xdsl/dialects/experimental/dmp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (georgebisbas) [reply xdsl/dialects/experimental/dmp.py:225]: We should follow some rule on what is generic and what not.
e.g. exchangedeclaration is generic?
but this is implied to have halo?

**Review by georgebisbas — COMMENTED** [review]
I think this converges to be merged

File: xdsl/dialects/experimental/dmp.py (hunk 8 above)
Comment (AntonLydike) [reply xdsl/dialects/experimental/dmp.py:129]: Good catch! I guess *rank* is an overloaded term here. It should probably be `dims` to be less ambigous. I'll add a docstring too

File: xdsl/dialects/experimental/dmp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/dmp.py:225]: Okay, I'll add the "_with_halo" suffix back. This dialect is about to undergo drastic changes anyway!

**Review by georgebisbas — APPROVED** [review]
I think we are happy...lets proceed

**Pull request merged**